Failed tests (10381): TestReplicateRogueRemovedNode

The following test appears to have failed:

[#10381](https://circleci.com/gh/cockroachdb/cockroach/10381):

```
W1228 15:49:51.779876 977 multiraft/multiraft.go:1297  node 1 failed to send message to 3: multiraft/transport.go:157: unknown peer 3
W1228 15:49:51.788759 977 multiraft/multiraft.go:1297  node 2 failed to send message to 3: multiraft/transport.go:157: unknown peer 3
I1228 15:49:51.832610 977 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I1228 15:49:51.832793 977 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I1228 15:49:51.832925 977 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestReplicateRogueRemovedNode (1.46s)
	<autogenerated>:32: storage/client_raft_test.go:1321: condition failed to evaluate within 1s: storage/client_test.go:575: expected [16 0 5], got [16 5 5]
=== RUN   TestReplicateReAddAfterDown
E1228 15:49:51.835017 977 gossip/gossip.go:180  different node IDs were set for the same gossip instance (2147483647, 1)
I1228 15:49:51.836157 977 storage/replica.go:1307  gossiping cluster id  from store 1, range 1
I1228 15:49:51.836546 977 multiraft/multiraft.go:578  node 1 starting
I1228 15:49:51.837495 977 raft/raft.go:446  [group 1] 1 became follower at term 5
I1228 15:49:51.837604 977 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I1228 15:49:51.837683 977 multiraft/multiraft.go:998  node 1 campaigning because initial confstate is [1]
I1228 15:49:51.837727 977 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I1228 15:49:51.837774 977 raft/raft.go:459  [group 1] 1 became candidate at term 6
--
I1228 15:50:03.929918 977 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I1228 15:50:03.930140 977 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestLeaderAfterSplit (0.35s)
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.79s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	34.112s
=== RUN   TestBatchBasics
I1228 15:49:40.422118 1039 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I1228 15:49:40.425449 1039 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I1228 15:49:40.431756 1039 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -705,7 +705,7 @@ func TestStoreRangeDownReplicate(t *testing.T) {
 			kickedOffReplicationQueue := false
 			for _, store := range mtc.stores {
 				if _, ok := idSet[store.StoreID()]; !ok {
-					store.ForceReplicaGCScan(t)
+					store.ForceReplicaGCScanAndProcess(t)
 				} else if !kickedOffReplicationQueue {
 					store.ForceReplicationScan(t)
 					kickedOffReplicationQueue = true
@@ -905,7 +905,7 @@ func TestReplicateAddAndRemove(t *testing.T) {
 
 		// Wait out the leader lease and the unleased duration to make the replica GC'able.
 		mtc.manualClock.Increment(int64(storage.ReplicaGCQueueInactivityThreshold+storage.DefaultLeaderLeaseDuration) + 1)
-		mtc.stores[1].ForceReplicaGCScan(t)
+		mtc.stores[1].ForceReplicaGCScanAndProcess(t)
 
 		// The removed store no longer has any of the data from the range.
 		verify([]int64{39, 0, 39, 39})
@@ -1322,12 +1322,23 @@ func TestReplicateRogueRemovedNode(t *testing.T) {
 	}
 
 	// Wait for the replica to be GC'd on node 1.
-	mtc.manualClock.Increment(int64(storage.DefaultLeaderLeaseDuration+
-		storage.ReplicaGCQueueInactivityThreshold) + 1)
-	mtc.stores[1].ForceReplicaGCScan(t)
-
 	// Store 0 has two writes, 1 has erased everything, and 2 still has the first write.
-	mtc.waitForValues(roachpb.Key("a"), time.Second, []int64{16, 0, 5})
+	// A single pass of ForceReplicaGCScanAndProcess is not enough, since the replica
+	// may be recreated by a stray raft message, so we run the GC scan inside the loop.
+	// TODO(bdarnell): if the call to RemoveReplica in replicaGCQueue.process can be
+	// moved under the lock, then the GC scan can be moved out of this loop.
+	util.SucceedsWithin(t, time.Second, func() error {
+		mtc.manualClock.Increment(int64(storage.DefaultLeaderLeaseDuration+
+			storage.ReplicaGCQueueInactivityThreshold) + 1)
+		mtc.stores[1].ForceReplicaGCScanAndProcess(t)
+
+		actual := mtc.readIntFromEngines(roachpb.Key("a"))
+		expected := []int64{16, 0, 5}
+		if !reflect.DeepEqual(expected, actual) {
+			return util.Errorf("expected %v, got %v", expected, actual)
+		}
+		return nil
+	})
 
 	// Bring node 2 back up.
 	mtc.restartStore(2)
@@ -1365,7 +1376,7 @@ func TestReplicateRogueRemovedNode(t *testing.T) {
 	// lease will cause GC to do a consistent range lookup, where it
 	// will see that the range has been moved and delete the old
 	// replica.
-	mtc.stores[2].ForceReplicaGCScan(t)
+	mtc.stores[2].ForceReplicaGCScanAndProcess(t)
 	mtc.waitForValues(roachpb.Key("a"), 3*time.Second, []int64{16, 0, 0})
 
 	// Now that the group has been GC'd, the goroutine that was

--- a/storage/client_replica_gc_test.go
+++ b/storage/client_replica_gc_test.go
@@ -113,7 +113,7 @@ func TestReplicaGCQueueDropReplicaGCOnScan(t *testing.T) {
 	// Make sure the range is removed from the store.
 	util.SucceedsWithin(t, time.Second, func() error {
 		store := mtc.stores[1]
-		store.ForceReplicaGCScan(t)
+		store.ForceReplicaGCScanAndProcess(t)
 		if _, err := store.GetReplica(rangeID); !testutils.IsError(err, "range .* was not found") {
 			return util.Errorf("expected range removal: %s", err)
 		}

--- a/storage/queue.go
+++ b/storage/queue.go
@@ -333,6 +333,7 @@ func (bq *baseQueue) processOne(clock *hlc.Clock) {
 
 // processReplica processes a single replica.
 // This should not be called externally to the queue.
+// bq.Lock should not be held while calling this method.
 func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) {
 	start := time.Now()
 	now := clock.Now()
@@ -400,8 +401,12 @@ func (bq *baseQueue) remove(index int) {
 // Exposed for testing only.
 func (bq *baseQueue) DrainQueue(clock *hlc.Clock) {
 	bq.Lock()
-	defer bq.Unlock()
-	for repl := bq.pop(); repl != nil; repl = bq.pop() {
+	repl := bq.pop()
+	bq.Unlock()
+	for repl != nil {
 		bq.processReplica(repl, clock)
+		bq.Lock()
+		repl = bq.pop()
+		bq.Unlock()
 	}
 }

--- a/storage/store.go
+++ b/storage/store.go
@@ -739,15 +739,17 @@ func (s *Store) ForceReplicationScan(t util.Tester) {
 	}
 }
 
-// ForceReplicaGCScan iterates over all ranges and enqueues any that
+// ForceReplicaGCScanAndProcess iterates over all ranges and enqueues any that
 // may need to be GC'd. Exposed only for testing.
-func (s *Store) ForceReplicaGCScan(t util.Tester) {
+func (s *Store) ForceReplicaGCScanAndProcess(t util.Tester) {
 	s.mu.Lock()
-	defer s.mu.Unlock()
 
 	for _, r := range s.replicas {
 		s.replicaGCQueue.MaybeAdd(r, s.ctx.Clock.Now())
 	}
+	s.mu.Unlock()
+
+	s.replicaGCQueue.DrainQueue(s.ctx.Clock)
 }
 
 // DisableRaftLogQueue disables or enables the raft log queue.

